hudi-utilities/src/test/java/org/apache/hudi/utilities/streamer/TestFetchNextBatchFromSource.java

```diff
@@ -0,0 +1,199 @@
+    when(sourceFormatAdapter.fetchNewDataInRowFormat(any(), anyLong())).thenReturn(fakeRowInputBatch);
+    //batch is empty because we don't want getBatch().map() to do anything because it calls static method we can't mock
+    InputBatch<JavaRDD<GenericRecord>> fakeAvroInputBatch = new InputBatch<>(Option.empty(), "chkpt", inputBatchSchemaProvider);
+    when(sourceFormatAdapter.fetchNewDataInAvroFormat(any(),anyLong())).thenReturn(fakeAvroInputBatch);
+
+    //transformer
+    //return empty because we don't want .map() to do anything because it calls static method we can't mock
+    when(sourceFormatAdapter.processErrorEvents(any(), any())).thenReturn(Option.empty());
+    Option<Transformer> transformerOption = Option.empty();
+    if (hasTransformer) {
+      transformerOption = Option.of(mock(Transformer.class));
+    }
+
+    //user provided schema provider
+    SchemaProvider schemaProvider = null;
+    if (hasSchemaProvider) {
+      schemaProvider = getSchemaProvider("UserProvided", nullTargetSchema);
+    }
+
+    //error table
+    TypedProperties props = new TypedProperties();
+    props.put(DataSourceWriteOptions.RECONCILE_SCHEMA().key(), false);
+    Option<BaseErrorTableWriter> errorTableWriterOption = Option.empty();
+    if (hasErrorTable) {
+      errorTableWriterOption = Option.of(mock(BaseErrorTableWriter.class));
+      props.put(ERROR_ENABLE_VALIDATE_TARGET_SCHEMA.key(), true);
+    }
+    TypedProperties propsSpy = spy(props);
+
+
+    //Actually create the deltastreamer
+    StreamSync streamSync = new StreamSync(cfg, sparkSession, propsSpy, hoodieSparkEngineContext,
+        fs, configuration, client -> true, schemaProvider, errorTableWriterOption, sourceFormatAdapter, transformerOption, useRowWriter, false);
+    StreamSync spy = spy(streamSync);
+    SchemaProvider deducedSchemaProvider;
+    if (deduceSchemaSame) {
+      deducedSchemaProvider = inputBatchSchemaProvider;
+    } else {
+      deducedSchemaProvider = getSchemaProvider("deduced", false);
+    }
+    doReturn(deducedSchemaProvider).when(spy).getDeducedSchemaProvider(any(), any(), any());
+
+    //run the method we are unit testing:
+    spy.fetchNextBatchFromSource(Option.empty(), mock(HoodieTableMetaClient.class));
+
+    //make sure getDeducedSchemaProvider is always called once
+    verify(spy, times(1)).getDeducedSchemaProvider(any(), any(), any());
+
+    //make sure we use error table when we should
+    //
+    //When should we use error table?
+    // - error table enabled obviously: hasErrorTable
+    // - row writer disabled (not implemented yet): !useRowWriter
+    // - if there is a transformer: hasTransformer
+    //    - if schema provider is present and the target schema is non null: !nullTargetSchema && hasSchemaProvider
+    // - if there is no transformer: !hasTransformer
+    //    - if the deduced schema is different than the schema of the input batch:  !deduceSchemaSame
+    verify(propsSpy, hasErrorTable && !useRowWriter
```

Comment:
this is hard to read.
I would suggest to add an argument to testCases only on whats expected. 
whether we expect deduceWriterSchema to be called or not. 
and whats the final writer schema that we expect. 
whether is target schema or the table schema or some deduced schema. 
if we can argument this, then assertion here would be very lean and easier to read.

Reply:
fixed